File: core/src/main/java/hello/core/AppConfig.java
package hello.core;

import hello.core.discount.DiscountPolicy;
import hello.core.discount.implement.FixDiscountPolicy;
import hello.core.discount.implement.RateDiscountPolicy;
import hello.core.member.repository.MemberRepository;
import hello.core.member.repository.MemberRepositoryImpl;
import hello.core.member.service.MemberService;
import hello.core.member.service.MemberServiceImpl;
import hello.core.order.implement.OrderServiceImpl;
import hello.core.order.service.OrderService;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import javax.sound.midi.Soundbank;

@Configuration
public class AppConfig {

    /**
     * 아래와 같이 Refactoring 하는게 좋은 이유
     * 1. Method 이름으로 역할이 모두 드러난다.
     * 2. Repo를 지금처럼 Memory가 아니라, DB를 사용한다고 하면, 해당 클래스 명칭만 바꾸면 된다.
     */

    // @Bean은 memberService를 호출하면서 new MemberRepositoryImpl()를 한 번 호출 해줌
    // @Bean은 OrderService를 호출하면서 new MemberRepositoryImpl()를 한 번 호출 해줌
    // 결과적으로 MemberRepositoryImpl를 총 2번 호출하게 됨.

    @Bean
    public MemberService memberService() {
        System.out.println("call AppConfig.memberService");
        return new MemberServiceImpl(memberRepository());
    }

    @Bean
    public OrderService orderService() {
        System.out.println("call AppConfig.orderService");
        return new OrderServiceImpl(
                memberRepository(),
                discountPolicy()
        );
    }

    @Bean
    public MemberRepository memberRepository() {
        System.out.println("call AppConfig.memberRepository");
        return new MemberRepositoryImpl();
    }

    @Bean
    public DiscountPolicy discountPolicy() {
//        System.out.println("call AppConfig.discountPolicy");
        return new RateDiscountPolicy();
    }
}

File: core-kotlin/src/main/kotlin/hello/corekotlin/AppConfig.kt
package hello.corekotlin

import hello.corekotlin.discount.DiscountPolicy
import hello.corekotlin.discount.implement.RateDiscountPolicy
import hello.corekotlin.member.repository.MemberRepository
import hello.corekotlin.member.repository.MemoryMemberImpl
import hello.corekotlin.member.service.MemberService
import hello.corekotlin.member.service.MemberServiceImpl
import hello.corekotlin.order.implement.OrderServiceImpl
import hello.corekotlin.order.service.OrderService
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class AppConfig {

    @Bean
    fun memberService(): MemberService = MemberServiceImpl(memberRepository())

    @Bean
    fun orderService(): OrderService = OrderServiceImpl(
        memberRepository(),
        discountPolicy()
    )

    @Bean
    fun memberRepository(): MemberRepository = MemoryMemberImpl()

    @Bean
    fun discountPolicy(): DiscountPolicy = RateDiscountPolicy()

}